Update `oc idle` to patch dc's and rc's instead of update

juanvallejo opened the pull request:
Fixes https://github.com/openshift/origin/issues/11622 as well as a minor bug when running the command without `--dry-run`
cc @openshift/cli-review @kargakis

juanvallejo commented:
[test]

kargakis commented:
@mfojtik i would like to have this for 1.4

juanvallejo commented on pkg/unidling/controller/controller.go:
```diff
@@ -325,7 +326,8 @@ func (c *UnidlingController) handleRequest(info types.NamespacedName, lastFired
 		scale.Spec.Replicas = scalableRef.Replicas
 
-		if err = scaleAnnotater.UpdateObjectScale(info.Namespace, scalableRef.CrossGroupObjectReference, obj, scale); err != nil {
+		codec := kapi.Codecs.LegacyCodec(registered.EnabledVersions()...)
```
@liggitt Switched to use `runtime.Encode()`, however I am not sure if this is the right way of obtaining a codec here

juanvallejo commented:
Hm, when I use a runtime encoder `cmdutil.Factory.JSONEncoder()` instead of just marshaling the object with the `json` package, I receive an error when running `oc idle` on an endpoint:

```
$ oc idle docker-registry
Marked service default/docker-registry to unidle resource DeploymentConfig default/docker-registry
(unidle to 1 replicas)
error: unable to scale DeploymentConfig default/docker-registry to 0, but still listed as target for unidling: unable to find api field in struct DeploymentConfig for the json field "metadata"
```
deploymentconfig object when using `json.Marshal`:

```
{"name":"docker-registry","namespace":"default","selfLink":"/oapi/v1/namespaces/default/deploymentconfigs/docker-registry","uid":"19fdb36f-9d51-11e6-9e27-507b9dac96e1","resourceVersion":"588","generation":1,"creationTimestamp":"2016-10-28T20:57:13Z","labels":{"docker-registry":"default"},"annotations":{"idling.alpha.openshift.io/idled-at":"2016-10-28T21:26:16Z","idling.alpha.openshift.io/previous-scale":"1"},"Spec":{"Strategy":{"Type":"Rolling","CustomParams":null,"RecreateParams":null,"RollingParams":{"UpdatePeriodSeconds":1,"IntervalSeconds":1,"TimeoutSeconds":600,"MaxUnavailable":"25%","MaxSurge":"25%","Pre":null,"Post":null},"Resources":{},"Labels":null,"Annotations":null},"MinReadySeconds":0,"Triggers":[{"Type":"ConfigChange","ImageChangeParams":null}],"Replicas":0,"RevisionHistoryLimit":null,"Test":false,"Paused":false,"Selector":{"docker-registry":"default"},"Template":{"metadata":{"creationTimestamp":null,"labels":{"docker-registry":"default"}},"spec":{"volumes":[{"name":"registry-storage","emptyDir":{}}],"containers":[{"name":"registry","image":"openshift/origin-docker-registry:v1.3.0-rc1","ports":[{"containerPort":5000,"protocol":"TCP"}],"env":[{"name":"REGISTRY_HTTP_ADDR","value":":5000"},{"name":"REGISTRY_HTTP_NET","value":"tcp"},{"name":"REGISTRY_HTTP_SECRET","value":"3hlMyRNRzcGziBH4tFIBH8RVvnyX5q6BLmG0rFpPZZ0="},{"name":"REGISTRY_MIDDLEWARE_REPOSITORY_OPENSHIFT_ENFORCEQUOTA","value":"false"}],"resources":{"requests":{"cpu":"100m","memory":"256Mi"}},"volumeMounts":[{"name":"registry-storage","mountPath":"/registry"}],"livenessProbe":{"httpGet":{"path":"/healthz","port":5000,"scheme":"HTTP"},"initialDelaySeconds":10,"timeoutSeconds":5,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"readinessProbe":{"httpGet":{"path":"/healthz","port":5000,"scheme":"HTTP"},"timeoutSeconds":5,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent","securityContext":{"privileged":false}}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"registry","securityContext":{}}}},"Status":{"LatestVersion":1,"ObservedGeneration":1,"Replicas":1,"UpdatedReplicas":1,"AvailableReplicas":1,"UnavailableReplicas":0,"Details":{"Message":"caused by a config change","Causes":[{"Type":"ConfigChange","ImageTrigger":null}]},"Conditions":null}}
```

same object when using `f.JSONEncoder()`:
```
{"kind":"DeploymentConfig","apiVersion":"v1","metadata":{"name":"docker-registry","namespace":"default","selfLink":"/oapi/v1/namespaces/default/deploymentconfigs/docker-registry","uid":"19fdb36f-9d51-11e6-9e27-507b9dac96e1","resourceVersion":"588","generation":1,"creationTimestamp":"2016-10-28T20:57:13Z","labels":{"docker-registry":"default"}},"spec":{"strategy":{"type":"Rolling","rollingParams":{"updatePeriodSeconds":1,"intervalSeconds":1,"timeoutSeconds":600,"maxUnavailable":"25%","maxSurge":"25%"},"resources":{}},"triggers":[{"type":"ConfigChange"}],"replicas":1,"test":false,"selector":{"docker-registry":"default"},"template":{"metadata":{"creationTimestamp":null,"labels":{"docker-registry":"default"}},"spec":{"volumes":[{"name":"registry-storage","emptyDir":{}}],"containers":[{"name":"registry","image":"openshift/origin-docker-registry:v1.3.0-rc1","ports":[{"containerPort":5000,"protocol":"TCP"}],"env":[{"name":"REGISTRY_HTTP_ADDR","value":":5000"},{"name":"REGISTRY_HTTP_NET","value":"tcp"},{"name":"REGISTRY_HTTP_SECRET","value":"3hlMyRNRzcGziBH4tFIBH8RVvnyX5q6BLmG0rFpPZZ0="},{"name":"REGISTRY_MIDDLEWARE_REPOSITORY_OPENSHIFT_ENFORCEQUOTA","value":"false"}],"resources":{"requests":{"cpu":"100m","memory":"256Mi"}},"volumeMounts":[{"name":"registry-storage","mountPath":"/registry"}],"livenessProbe":{"httpGet":{"path":"/healthz","port":5000,"scheme":"HTTP"},"initialDelaySeconds":10,"timeoutSeconds":5,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"readinessProbe":{"httpGet":{"path":"/healthz","port":5000,"scheme":"HTTP"},"timeoutSeconds":5,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent","securityContext":{"privileged":false}}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"registry","serviceAccount":"registry","securityContext":{}}}},"status":{"latestVersion":1,"observedGeneration":1,"replicas":1,"updatedReplicas":1,"availableReplicas":1,"details":{"message":"caused by a config change","causes":[{"type":"ConfigChange"}]}}}
```

kargakis commented:
Also fixes https://github.com/openshift/origin/issues/10988

juanvallejo commented:
cc @DirectXMan12 @fabianofranz 

Per Solly's suggestion, I have added a `ScaleUpdater` interface with an `Update()` method that allows to update or patch a resource depending on what is needed. It was not seen as a good idea to patch resources [when calling `UpdateObjectScale() from `unidling/controller/controller.go`](https://github.com/openshift/origin/blob/master/pkg/unidling/controller/controller.go#L328), so the implementation with `Patch`ing is only seen in `cmd/idle.go`

fabianofranz commented:
Win, now this is looking like properly architected. I imagine this is already covered by tests so we don't need to write new ones, correct?
I'll defer to @DirectXMan12 a final review.

smarterclayton commented:
Can I get a quick summary of why we can't use patch in scale?  Do we need
precondition scale?

On Nov 1, 2016, at 9:46 AM, Fabiano Franz <notifications@github.com> wrote:

Win, now this is looking like properly architected. I imagine this is
already covered by tests so we don't need to write new ones, correct?
I'll defer to @DirectXMan12 <https://github.com/DirectXMan12> a final
review.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11634#issuecomment-257569936>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_pzZneCgEspPFhlpt8bb1jcazvfs2ks5q50KYgaJpZM4Ki1Nq>
.

liggitt commented:
Patch on scale with name/namespace/uid/resourceVersion set should work correctly.

smarterclayton commented:
I think this is patch under scale - an unrelated change should not block
scaling.  Probably we should be doing scale with no resource version
precondition

On Nov 1, 2016, at 9:56 AM, Jordan Liggitt <notifications@github.com> wrote:

Patch on scale with name/namespace/uid/resourceVersion set should work
correctly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11634#issuecomment-257572454>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p0Jd3dC3UlS8UrVqMGcrYlzH1hnYks5q50UdgaJpZM4Ki1Nq>
.

smarterclayton commented:
This is a problem with scale, not with the clients of Scale.  If this is a
blind scale, then we should be ignoring the existing value of scale, and we
should not be sending a resource version when we POST the scale resource,
and the scale resource on the server should use patch under the covers to
ignore what else the user provides.

On Tue, Nov 1, 2016 at 1:49 PM, Juan Vallejo <notifications@github.com>
wrote:

> @liggitt <https://github.com/liggitt> @smarterclayton
> <https://github.com/smarterclayton> Should I update unidling/controller/
> controller.go
> <https://github.com/openshift/origin/blob/master/pkg/unidling/controller/controller.go#L328>
> to patch as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11634#issuecomment-257638740>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6OVGwgZ7zTRCabxH1kf2wSwyE7qks5q53uogaJpZM4Ki1Nq>
> .
>

juanvallejo commented:
@smarterclayton @DirectXMan12 updated `unidling/controller/controller.go` to use `Patch()` as well

The only times Patch() is not used [is when the resource kind](https://github.com/openshift/origin/blob/master/pkg/unidling/util/scale.go#L84) [is unknown](https://github.com/openshift/origin/blob/master/pkg/unidling/util/scale.go#L105)

PTAL

juanvallejo commented:
@DirectXMan12 Thanks for the feedback! Updated `ScaleUpdater` to use its own `dcGetter` / `rcGetter` and reverted `ScaleAnnotater`'s fields back to "private"

kargakis commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -12,6 +12,7 @@ import (
 
 	cmdutil "github.com/openshift/origin/pkg/cmd/util"
 	utilerrors "github.com/openshift/origin/pkg/util/errors"
```
Move these imports with the rest of the Openshift imports

juanvallejo commented:
@kargakis 
> Move these imports with the rest of the Openshift imports

Done!

juanvallejo commented on pkg/unidling/controller/controller_test.go:
```diff
@@ -131,6 +131,40 @@ func prepFakeClient(t *testing.T, nowTime time.Time, scales ...kextapi.Scale) (*
 		return true, nil, errors.NewNotFound(action.GetResource().GroupResource(), obj.Name)
 	})
 
+	fakeDeployClient.PrependReactor("patch", "deploymentconfigs", func(action ktestingcore.Action) (bool, runtime.Object, error) {
```
@DirectXMan12 or @kargakis PTAL

soltysh commented on pkg/unidling/controller/controller.go:
```diff
@@ -61,6 +63,68 @@ func (c *lastFiredCache) AddIfNewer(info types.NamespacedName, newLastFired time
+		annotator.ChangeAnnotations(typedObj.Spec.Replicas, typedObj.Annotations)
+		typedObj.Spec.Replicas = scale.Spec.Replicas
+
+		newObj, err := json.Marshal(typedObj)
```
That wasn't still addressed, we should not be using raw json package for any kind of encoding. Sooner or later this will bite us hard. I'd rather fix it before that happens. Here and the next call in the following case, as well.

juanvallejo replied:
@soltysh
> That wasn't still addressed, we should not be using raw json package for any kind of encoding. Sooner or later this will bite us hard. I'd rather fix it before that happens. Here and the next call in the following case, as well.

Thanks for the feedback, done!

juanvallejo commented:
conformance flaked on https://github.com/openshift/origin/issues/10988 re[test]

fabianofranz commented:
flaked on https://github.com/openshift/origin/issues/9076 re[test]

juanvallejo commented:
conformance flaked on https://github.com/openshift/origin/issues/11775 re[test]

fabianofranz commented:
@DirectXMan12 @smarterclayton any other comments here?

juanvallejo commented:
@DirectXMan12 
> I'd prefer for the scaleUpdater to be moved to utils if it's being used by both the controller and oc

Makes sense, done

juanvallejo commented:
@DirectXMan12 @fabianofranz I will go ahead and squash if there are no more comments, thanks for the review

fabianofranz commented:
LGTM [merge]
